adr/ADR-42.md

```diff
@@ -0,0 +1,168 @@
+
+Related prior work:
+
+ * Proposed [Consumer Groups](https://github.com/nats-io/nats-architecture-and-design/pull/36)
+ * Proposed [Partitioned consumer groups and exclusive consumer](https://github.com/nats-io/nats-architecture-and-design/pull/263)
+ * Proposed [Consumer Owner ID](https://github.com/nats-io/nats-server/pull/5157)
+
+## General Overview
+
+We introduce 2 settings on `ConsumerConfig` that activates these related features:
+
+```go
+{
+	PriorityGroups: ["<group>", ...],
+	PriorityPolicy: "<policy>"
+}
+```
+
+The presence of the `PriorityGroups` setting activates the set of features documented here with a list of names supported
+behaviours configured using the `PriorityPolicy`.
+
+Technically in some of the proposed policies the `PriorityGroups` have no real meaning today, but we keep it for consistency 
+and allow for future features to be added that would be per-group without then requiring all clients to be updated. 
+Future message grouping features would require groups to be listed here.
+
+In the initial implementation we should limit `PriorityGroups` to one per consumer only and error should one be made with 
+multiple groups. In future iterations multiple groups will be supported along with dynamic partitioning of stream
+data.
+
+This is only supported on Pull Consumers, configuring this on a Push consumer must raise an error.
+
+## Priority Policies
+
+### `overflow` policy
+
+Users want certain clients to only get messages when certain criteria are met.
+
+Imagine jobs are best processed locally in `us-east-1` but at times there might be so many jobs in that region that 
+`us-west-1` region handling some overflow, while less optimal in terms of transit costs and latency, would be desirable 
+to ensure serving client needs as soon as possible.
+
+The `overflow` policy enables Pull requests to be served only if criteria like `num_pending` and `num_ack_pending` are
+above a certain limit, otherwise those Pull requests will sit idle in the same way that they would if no messages were
+available (receiving heartbeats etc).
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "overflow",
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+Here we state that the Consumer has one group called `jobs`, it is operating on the `overflow` policy and requires `explicit`
+Acks, any other Ack policy will produce an error. If we force this ack policy in normal use we should error in Pedantic mode.
+
+Pull requests will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"min_pending": 1000` - only deliver messages when `num_pending` for the consumer is >= 1000
+ * `"min_ack_pending: 1000` - only deliver messages when `ack_pending` for the consumer is >= 1000
+
+If `min_pending` and `min_ack_pending` are both given both have to be satisfied (boolean AND).
```

Comment:
I wonder if this is better, or the way how stream limits works - first threshold passed means overflow condition is satisfied (boolean OR).

I think that in most cases users want threshold condition to kick in, so OR makes more sense, however I can clerly understand that AND could also be useful.

We could reason about it this way:
- OR - we want a looser conditions, that allow overflowing if processing clients are too slow, or they try to process too much at once (max ack pending is high)
- AND - we want be able to tighten the conditions and only overflow if both - clients process too fast/with too much parallelism and are lagging behind.

I think that the second case is rare. Either clients are not catching up, or are trying "too hard" (depending on code and user strategy). Matching both criterias sounds like a "approaching imminent disaster" scenario, where overlow tries to prevent that. Which, when I just said it, actually makes an interesting use case...

Reply:
I think you're right yeah, will update